Cargo lib bin split

dsheets opened the pull request:
Hi! I wrote this exact module a few years ago in python for a Christmas present but I'm now rewriting that present (in rust because python makes me sad) to integrate more features for another Christmas present and I came across this repo. Thanks so much for publishing it! You're saving me hours of time re-implementing.

Anyway, I would really like to have the executable and the library separate so I've split them. It builds now and the tests pass and clippy is silent.

I'm happy to collaborate to get this merged. Thanks!

dsheets committed:
Split into lib (spotify-connect-client) and executable crates

dsheets committed:
Add *~ to .gitignore

dsheets committed:
Silencio el clippy!

TimotheeGerber commented:
Hi @dsheets!

Making both a lib crate and a bin crate is something I had in mind since I started this project. So your MR is more than welcome! Thank you for taking the time to do it.

I'll have a look to your code this week and let you know if I can merge it as is or if some modifications are needed.

dsheets committed:
net: fix an addUser method/querystring/body bug

Key-value pairs were being POSTed (correct by Spotify's spec) for
action=addUser but they were in the query string rather than in the
x-www-form-urlencoded body. This caused librespot-java 1.6.3 to refuse
connection as it could not find the action type (addUser) because it
wasn't in the body. The official Spotify client uses the body of the
POST. The rust librespot application must be lenient in this regard.

The added urlencoding dependency was already a transitive dependency
of the minreq create with 'urlencoding' feature.

dsheets committed:
main: split authenticate function out of main

Also adds log and env_logger dependencies so the broken out function
can report its status.

dsheets commented:
Great, thanks! I fixed a bug I found when trying to connect to librespot-java and I started extracting more functionality from the binary into the library.

dsheets committed:
Move main::authenticate to the top level of the library

Also clean up a few dependencies that the binary had that were
transitive dependencies only

dsheets committed:
main: loosen some dependency version constraints

dsheets committed:
net::DeviceInfo: cleanup, comments, add activeUser field

Also make the type serde Serializable because one may wish to store
device info to examine later

dsheets committed:
main: tell the user who was booted from the device (if any)

dsheets committed:
authenticate: accept hostnames, too

dsheets committed:
Update librespot dependencies to 0.4.2

dsheets committed:
auth: return errors instead of panicking

TimotheeGerber commented on lib/src/net.rs:
```diff
@@ -1,21 +1,20 @@
-    #[serde(rename = "clientID")]
-    pub client_id: Option<String>,
-    pub scope: Option<String>,
+    pub active_user: Option<String>, // undocumented but useful and returned by both librespot and librespot-java
```
I tried both with an old 0.2.0 version and the latest 0.4.2 version of `librespot`. In each case, I got `Some("")` for the `activeUser` field. Does it work better on `librespot-java`?

dsheets replied:
That is the expected behavior if the field is there but no user is currently connected. I did have to work around that for my use case but also it would be good to be able to tell if the field is missing/unsupported or if there is just no user connected right now. Ideas?

TimotheeGerber replied:
I should have been more explicit. Even when connected and streaming music from `librespot`, the returned `activeUser` is empty. Which seems not the desired behavior. It's harmless, so I'll keep it for now, especially if it works well for your use case with `librespot-java`.

Concerning the distinction between the missing field and the empty field, I am not sure it is currently possible while using the `serde` to deserialize the `DeviceInfo` struct. It could be distinguished if the struct were deserialized manually, but that's kind of verbose, error prone and not very idiomatic.

dsheets replied:
Oh, that's odd. It is definitely supposed to support it -- see librespot-org/librespot#1235 for the beginning of a fix.

TimotheeGerber replied:
Really nice follow up! Thanks for taking so much time improving code, whatever the repository.

I tested your changes on `librespot`. It works perfectly and integrates nicely with the `spotify-connect` command. Well done!

TimotheeGerber reviewed (commented):
Wow! You did an incredible work. Not only you separate the project in a lib and a bin crates, but you also add proper error handling, cleaned/updated dependencies and added some new features. Thanks a lot for this huge contribution!

I tested it on my local network and it still works as previously.

I only have a change request about the `clientID`. I can not test it because I don't have an official Spotify hardware, but I'm pretty sure it is important. Once corrected, I will merge your contribution.

The small question about `activeUser` is not blocking. I am just curious.

dsheets committed:
lib::net: fix clientID serialized key name

dsheets commented on lib/src/net.rs:
```diff
@@ -1,21 +1,20 @@
-    pub scope: Option<String>,
+    pub active_user: Option<String>, // undocumented but useful and returned by both librespot and librespot-java
+    pub token_type: Option<String>,  // required at least as of 2.9.0
+    pub client_id: Option<String>,   // required at least as of 2.9.0
```
9b0844a (I'm not sure why `rust-fmt` thinks the comments should be aligned like that ¯\_(ツ)_/¯)

TimotheeGerber commented:
Merged :tada: 

Once again, thank you for this major contribution. Good luck with your Christmas presents!
